aten/src/ATen/Storage.h

```diff
@@ -1,53 +1,122 @@
 #pragma once
 
-#include "ATen/Scalar.h"
-#include <TH/THStorageFunctions.hpp>
+#include <ATen/Scalar.h>
+
+#include <ATen/Allocator.h>
+#include <ATen/ScalarType.h>
+#include <ATen/ScalarTypeUtils.h>
+#include <TH/THTypeConversion.hpp>
+#include <atomic>
+
+// Note [Weak references for intrusive refcounting]
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+// Here's the scheme:
+//
+//  - refcount == number of strong references to the object
+//    weakcount == number of weak references to the object,
+//      plus one more if refcount > 0
+//
+//  - THStorage stays live as long as there are any strong
+//    or weak pointers to it (weakcount > 0, since strong
+//    references count as a +1 to weakcount)
+//
+//  - finalizers are called and data_ptr is deallocated when refcount == 0
+//
+//  - Once refcount == 0, it can never again be > 0 (the transition
+//    from > 0 to == 0 is monotonic)
+//
+//  - When you access THStorage via a weak pointer, you must
+//    atomically increment the use count, if it is greater than 0.
+//    If it is not, you must report that the storage is dead.
+//
+
+struct THFinalizer {
+  virtual void operator()() = 0;
+  virtual ~THFinalizer() {};
+};
 
 namespace at {
 
 struct Type;
 
 struct Storage {
-  static const char RESIZABLE = 2;
 
-  Storage() {}
-  Storage(THStorage* storage)
-      : storage(storage) {}
-  Storage(const Storage& other) = delete;
-  virtual ~Storage() {
-    THStorage_free(storage);
+  Storage() = delete;
+  virtual ~Storage() {};
+  Storage(at::Backend, at::ScalarType, ptrdiff_t, at::DataPtr, at::Allocator*, char);
+  Storage(at::Backend, at::ScalarType, ptrdiff_t, at::Allocator*, char);
```

Comment:
Nit: can you add spaces here, so it's easier to visually distinguish between constructors, members and functions?